File: sample/src/main/java/ru/terrakok/cicerone/sample/ui/animations/photos/SelectPhotoFragment.java
package ru.terrakok.cicerone.sample.ui.animations.photos;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.arellomobile.mvp.MvpAppCompatFragment;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;

import javax.inject.Inject;

import ru.terrakok.cicerone.Router;
import ru.terrakok.cicerone.sample.R;
import ru.terrakok.cicerone.sample.SampleApplication;
import ru.terrakok.cicerone.sample.mvp.animation.photos.SelectPhotoPresenter;
import ru.terrakok.cicerone.sample.mvp.animation.photos.SelectPhotoView;
import ru.terrakok.cicerone.sample.ui.animations.ProfileActivity;
import ru.terrakok.cicerone.sample.ui.common.BackButtonListener;

/**
 * Created by Konstantin Tskhovrebov (aka @terrakok) on 14.07.17.
 */

public class SelectPhotoFragment extends MvpAppCompatFragment implements SelectPhotoView, BackButtonListener {
    private static final String ARG_RESULT_CODE = "arg_result_code";
    private static final String ARG_ANIM_DESTINATION = "arg_anim_dest";

    private ImageView photo1;
    private ImageView photo2;
    private ImageView photo3;
    private ImageView photo4;

    @Inject
    Router router;

    @InjectPresenter
    SelectPhotoPresenter presenter;

    public static SelectPhotoFragment getNewInstance(int resultCode) {
        SelectPhotoFragment fragment = new SelectPhotoFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_RESULT_CODE, resultCode);
        fragment.setArguments(args);
        return fragment;
    }

    public void setAnimationDestinationId(int resId) {
        Bundle arguments = getArguments();
        arguments.putInt(ARG_ANIM_DESTINATION, resId);
        setArguments(arguments);
    }

    private int getAnimationDestionationId() {
        return getArguments().getInt(ARG_ANIM_DESTINATION);
    }

    @ProvidePresenter
    SelectPhotoPresenter providePresenter() {
        return new SelectPhotoPresenter(router, getArguments().getInt(ARG_RESULT_CODE));
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        SampleApplication.INSTANCE.getAppComponent().inject(this);
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_select_photo, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        photo1 = (ImageView) view.findViewById(R.id.select_image_1);
        photo2 = (ImageView) view.findViewById(R.id.select_image_2);
        photo3 = (ImageView) view.findViewById(R.id.select_image_3);
        photo4 = (ImageView) view.findViewById(R.id.select_image_4);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        photo1.setOnClickListener(clickListener);
        photo2.setOnClickListener(clickListener);
        photo3.setOnClickListener(clickListener);
        photo4.setOnClickListener(clickListener);
    }

    private View.OnClickListener clickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            photo1.setTransitionName(null);
            photo2.setTransitionName(null);
            photo3.setTransitionName(null);
            photo4.setTransitionName(null);
            v.setTransitionName(ProfileActivity.PHOTO_TRANSITION);
            presenter.onPhotoClick((Integer) v.getTag());
        }
    };

    @Override
    public void showPhotos(int[] resurceIds) {
        if (resurceIds.length >= 4) {
            photo1.setImageResource(resurceIds[0]);
            photo2.setImageResource(resurceIds[1]);
            photo3.setImageResource(resurceIds[2]);
            photo4.setImageResource(resurceIds[3]);

            photo1.setTag(resurceIds[0]);
            photo2.setTag(resurceIds[1]);
            photo3.setTag(resurceIds[2]);
            photo4.setTag(resurceIds[3]);

            //for shared element animation
            int animRes = getAnimationDestionationId();
            photo1.setTransitionName(animRes == resurceIds[0] ? ProfileActivity.PHOTO_TRANSITION : null);
            photo2.setTransitionName(animRes == resurceIds[1] ? ProfileActivity.PHOTO_TRANSITION : null);
            photo3.setTransitionName(animRes == resurceIds[2] ? ProfileActivity.PHOTO_TRANSITION : null);
            photo4.setTransitionName(animRes == resurceIds[3] ? ProfileActivity.PHOTO_TRANSITION : null);
        }
    }

    @Override
    public boolean onBackPressed() {
        presenter.onBackPressed();
        return true;
    }
}

File: sample/src/main/java/com/github/terrakok/cicerone/sample/ui/animations/photos/SelectPhotoFragment.kt
package com.github.terrakok.cicerone.sample.ui.animations.photos

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import com.github.terrakok.cicerone.Router
import com.github.terrakok.cicerone.sample.SampleApplication
import com.github.terrakok.cicerone.sample.databinding.FragmentSelectPhotoBinding
import com.github.terrakok.cicerone.sample.mvp.animation.PhotoSelection
import com.github.terrakok.cicerone.sample.mvp.animation.photos.SelectPhotoPresenter
import com.github.terrakok.cicerone.sample.mvp.animation.photos.SelectPhotoView
import com.github.terrakok.cicerone.sample.ui.animations.ProfileActivity
import com.github.terrakok.cicerone.sample.ui.common.BackButtonListener
import moxy.MvpAppCompatFragment
import moxy.presenter.InjectPresenter
import moxy.presenter.ProvidePresenter
import javax.inject.Inject

/**
 * Created by Konstantin Tskhovrebov (aka @terrakok) on 14.07.17.
 */
class SelectPhotoFragment : MvpAppCompatFragment(), SelectPhotoView, BackButtonListener {

    private lateinit var binding: FragmentSelectPhotoBinding
    private lateinit var allImages: Array<ImageView>

    @Inject
    lateinit var router: Router

    @Inject
    lateinit var photoSelection: PhotoSelection

    @InjectPresenter
    lateinit var presenter: SelectPhotoPresenter

    fun setAnimationDestinationId(resId: Int) {
        var arguments = arguments
        if (arguments == null) arguments = Bundle()
        arguments.putInt(ARG_ANIM_DESTINATION, resId)
        setArguments(arguments)
    }

    private val animationDestionationId: Int
        get() = arguments!!.getInt(ARG_ANIM_DESTINATION)

    @ProvidePresenter
    fun providePresenter() = SelectPhotoPresenter(photoSelection, router)

    override fun onCreate(savedInstanceState: Bundle?) {
        SampleApplication.INSTANCE.appComponent.inject(this)
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentSelectPhotoBinding.inflate(inflater, container, false)
        allImages = arrayOf(
                binding.selectImage1,
                binding.selectImage2,
                binding.selectImage3,
                binding.selectImage4
        )
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        allImages.forEach {
            it.setOnClickListener(clickListener)
        }
    }

    private val clickListener = View.OnClickListener { v ->
        allImages.forEach {
            it.transitionName = null
        }
        v.transitionName = ProfileActivity.PHOTO_TRANSITION
        presenter.onPhotoClick((v.tag as Int))
    }

    override fun showPhotos(resourceIds: IntArray) {
        if (resourceIds.size >= 4) {
            //for shared element animation
            val animRes = animationDestionationId

            allImages.forEachIndexed {
                index, imageView ->
                imageView.setImageResource(resourceIds[index])
                imageView.tag = resourceIds[index]
                imageView.transitionName = if (animRes == resourceIds[index]) ProfileActivity.PHOTO_TRANSITION else null
            }
        }
    }

    override fun onBackPressed(): Boolean {
        presenter.onBackPressed()
        return true
    }

    companion object {
        private const val ARG_ANIM_DESTINATION = "arg_anim_dest"
    }
}